### PR TITLE
Added GLU (gated linear unit)

### PR DESCRIPTION
**Pull request by colesbury**
From https://arxiv.org/abs/1612.08083

### CODE REVIEW

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -279,6 +279,36 @@ def __repr__(self):
         return self.__class__.__name__ + inplace_str
 
 
+class GLU(Module):
+    """Applies the gated linear unit function :math:`{GLU}(a, b)= a \otimes \sigma(b)`
+    where `a` is the first half of the input vector and `b` is the second half.
```
Comment (apaszke) [inline torch/nn/modules/activation.py]: I don't see a reason why do we force the input to be a single vector. The interface would be much more logical if we were to accept `a` and `b` separately, as in the equation that's shown here.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -437,6 +437,12 @@ def relu(input, inplace=False):
     return _functions.thnn.Threshold.apply(input, 0, 0, inplace)
 
 
+def glu(input, dim=-1):
+    if dim < 0:
+        dim += input.dim()
```
Comment (apaszke) [inline torch/nn/functional.py]: I don't think you can ignore checks in here, or we're going to give weird error message.
```python
input = torch.randn(2, 2)
glu(input, dim=-4)
```
will fail with an error that dim -2 is out of range (while -2 is in fact a valid value for dim).

**Comment by apaszke** [comment]
Also, functional interface should capitalize the name as well

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (colesbury) [inline torch/nn/modules/activation.py]: I thought about this, but it's less convenient in practice. Gated linear units are used on the outputs of convolutional layers (https://arxiv.org/abs/1612.08083, https://arxiv.org/abs/1705.03122, https://arxiv.org/abs/1705.06824). You implement this with something like:

```
m = nn.Conv1d(input_planes, output_planes * 2)
F.glu(m(input))
```

For efficiency reasons, you don't want to split the Conv1d into two separate convolutions. If you took in two separate inputs, you'd have to add splitting code everywhere before you call glu.

**Review by colesbury — COMMENTED** [review]
We don't capitalize the names in functional interfaces (see relu, elu, etc.)

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (apaszke) [inline torch/nn/modules/activation.py]: I can't see why can't you replace it with this:
```python
m = nn.Conv1d(input_planes, output_planes * 2)
conv_feat = m(input)
size = conv_feat.size(1) // 2
F.glu(conv_feat[:size], conv_feat[size:])
```

Comment (apaszke) [inline torch/nn/modules/activation.py]: Eh ok. I guess we can do it this way

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
this is now merged into master
